`GraphEdit`/`GraphNode` Overhaul

dawdle-deer opened the pull request:
This PR restructures the graph system for friendlier usage and greater flexibility.
- Introduce `GraphPort`, a generic `GraphElement` to which connections can be made.
- Split `GraphNode` into two nodes: `GraphNodeIndexed`, which inherits the current `GraphNode` behaviour of associating children with slots and generating+placing `GraphPort`s automatically, and `GraphNode`, a base class designed for the flexible creation of nodes that contain ports.
- Introduce `GraphConnection`, a resource representing connections between `GraphPort`s in a `GraphEdit`.

Resolves https://github.com/godotengine/godot-proposals/issues/12207 - read the description there for more details.

https://github.com/user-attachments/assets/e5105b7c-5745-4c90-81f0-74f33d58678f


Notably, this implementation does not overhaul `VisualShader` or `VisualShaderNode`. I believe a future PR that overhauls the visual shader system in parallel would be a good idea, but for this PR, changes will be relegated to `VisualShaderNodePlugin` and seek to maintain compatibility as much as possible.

TODO:
- [x] Fix ~~so many~~ regressions and bugs
- [x] Ensure back-compatibility with visual shaders
- [x] Documentation
- [x] Improve keyboard navigation and accessibility features
- [ ] ~~Compatibility~~
- [ ] ~~Associate slots with nested children of `GraphNodeIndexed`~~

Edit 7/20/25: With GraphPort being an independent node, tracking nested children with `GraphNodeIndexed` slots has proven somewhat unnecessary. This could be reintroduced in a separate PR, if someone has need of it.
Edit 8/5/25: Certain aspects of this PR break compatibility beyond what I can handle. I'm going forward with the assumption that this will fully break compat. GraphShaders will still work as expected.

dawdle-deer committed:
begin GraphEdit/GraphNode overhaul

dawdle-deer committed:
continue graph system revision

dawdle-deer committed:
move connection logic out of port resource

dawdle-deer committed:
fix properties and such

dawdle-deer committed:
fix slot tracking on indexed graph node

dawdle-deer committed:
fix property management some, cleanup

dawdle-deer committed:
fix many various graph system issues

themes, layout, mouse interaction, connections, signals, etc

dawdle-deer committed:
expose GraphPort.on_disabled_behaviour

dawdle-deer committed:
progress towards fixing the visual shader editor

dawdle-deer commented on scene/gui/graph_edit.h:
```diff
@@ -406,33 +367,41 @@ class GraphEdit : public Control {
+	void move_connections(const Ref<GraphPort> p_from_port, const Ref<GraphPort> p_to_port);
+	bool is_connected(const Ref<GraphPort> p_first_port, const Ref<GraphPort> p_second_port);
+	int get_connection_count(const Ref<GraphPort> p_port);
+	GraphNode *get_connection_target(const Ref<GraphPort> p_port);
```
The method `get_connections_by_port` already serves this purpose, returning a `TypedArray` of all `GraphConnection`s on a given port.

You're right that this method should be changed, but currently, it's only used in `GraphNode` for keyboard navigation, so fixing this will wait until I rework the keyboard accessibility features of `GraphNode` and `GraphPort`.

Naros replied:
I will say other than this, I am really stoked at seeing these new APIs. We've had to build a lot of this as part of our own `GraphNode` implementations, so seeing this pulled up into the base implementation really helps to reduce a technical debt. And when you pair that with the idea of decoupling slot/index from ports and leaving that as an internal detail -- this is a huge improvement IMO; so definitely looking forward to this!

dawdle-deer replied:
Still need to rework keyboard nav, but I've added a new method `get_connected_ports` to `GraphEdit`, which returns an array of all ports connected to a given port. Should cover this use case!

dawdle-deer committed:
convert GraphPort to a GraphElement node

dawdle-deer committed:
fix port hotzones

dawdle-deer committed:
improve port layout, fix bugs

dawdle-deer committed:
Merge branch 'master' of https://github.com/dawdle-deer/godot into graphnode-revision

# Conflicts:
#	editor/shader/visual_shader_editor_plugin.cpp
#	editor/themes/editor_theme_manager.cpp
#	scene/gui/graph_edit.cpp
#	scene/gui/graph_node.cpp
#	scene/gui/graph_node.h

dawdle-deer committed:
incorporate https://github.com/godotengine/godot/pull/108245

dawdle-deer committed:
fix header class references

dawdle-deer committed:
fix errors

dawdle-deer committed:
fix small issues

dawdle-deer committed:
fix port colors, other bugs

dawdle-deer committed:
*finally* fix port+slot layout, fix name conflicts and bugs

dawdle-deer committed:
fix port expansion, get_filtered_port_count

dawdle-deer committed:
fix connections, port hotzones, many other issues

visual shaders are almost back to working as intended!! :D

dawdle-deer committed:
update API

dawdle-deer committed:
fix issue introduced by a prior fix

yo dawg i heard you like fixes, so we put fixes in your fixes so you can fix while you fix

dawdle-deer committed:
improve APIs and fix various issues

this commit marks a tipping point: i've begun rewriting VNML to rely on this PR! VNML was the project that justified this change to begin with, and now, changes will be made in tandem with testing to speed up bug-squashing and feature/API improvements

dawdle-deer committed:
add get_filtered_connections_by_node, update APIs

new method for GraphEdit that returns all connections to ports that match a filter direction on a node. also comes with similar methods on GraphNode and GraphNodeIndexed

dawdle-deer committed:
add more handy methods

mostly ease of use/shortcut methods on GraphNode and GraphPort, as well as get_first_connection_by_port on GraphEdit to shortcut get_connections_by_port()[0]

dawdle-deer committed:
further API handiness

dawdle-deer committed:
add yet more juicy juicy helper methods

this time, it's some ways to get all ports on a node that match a filter direction

dawdle-deer committed:
small fix for api

dawdle-deer committed:
begin keyboard accessibility update, add helpers

dawdle-deer committed:
fix for animation blend tree editor

dawdle-deer committed:
fix small issue

dawdle-deer committed:
expose port container node, fix edge cases and bugs

dawdle-deer committed:
improve GraphNode properties, fix port container, fix port layout

dawdle-deer committed:
rewrite slot indexing and data structure, fix slot reordering

dawdle-deer committed:
add slot signals

dawdle-deer committed:
move some callable initializations around

dawdle-deer committed:
fix and improve various things

dawdle-deer committed:
fix typo :P

dawdle-deer committed:
fix minor issue

dawdle-deer committed:
remove unused variables

dawdle-deer committed:
fix smol errors lol

dawdle-deer committed:
improve error handling slightly

dawdle-deer committed:
fix slot indexing for mixed heirarchies, handle edge cases

dawdle-deer committed:
fix signal typing

hey, why is PROPERTY_HINT_RESOURCE_TYPE the correct hint for Node properties of signals instead of PROPERTY_HINT_NODE_TYPE?

dawdle-deer committed:
update API

make GraphEdit::connect_nodes return the newly-created GraphConnection

dawdle-deer committed:
more API changes

very, *very* glad other people aren't relying on this yet, with how much testing has warranted API changes

dawdle-deer committed:
handle another edge case

dawdle-deer committed:
another helper method!

dawdle-deer committed:
handle another edge case

dawdle-deer committed:
add another handy method and some bounds checks

dawdle-deer committed:
add yet another helper method

dawdle-deer committed:
expose method i forgor to expose lol

also throw a check in there just in case

dawdle-deer committed:
change APIs slightly

dawdle-deer committed:
fix some issues

make mouse behaviour ignore disabled ports lol

dawdle-deer committed:
rework GraphNode APIs, move port container, add ignore_node_meta_tag

this mostly shuffles stuff between GraphNode and GraphNodeIndexed to put things in better places.
port_container is too opinionated for the base class, so it's been moved to GraphPortIndexed.
conversely, many of the input/output methods were simply shorthand for filtered methods, which are not as opinionated, so they've been moved to GraphNode.
ignore_node_meta_tag is a more generic way to mark a node to be ignored by GraphNode::_resort and GraphNodeIndexed slot associations. previously, this behaviour was only available to port_container and internal nodes, but this tag opens up the user to mark special nodes that ignore GraphNode rules

dawdle-deer committed:
lol

dawdle-deer committed:
update APIs, fix things

API changes aside, many of the functional changes here are meant to get GraphNodeIndexed to play nice with UndoRedo operations, such as by no longer freeing GraphNodeIndexed ports on slot removal

dawdle-deer committed:
oh lol most get_other methods were backwards

dawdle-deer committed:
improve exclusive ports & connection line data, add const, fix bugs

disconnection requests for exclusive ports are now emitted on _add_connection. also spam const everywhere :)

dawdle-deer committed:
update APIs, fix things

dawdle-deer committed:
Document it all !! also small fixes

First pass on documentation

dawdle-deer committed:
improve keyboard navigation some, fix node leaks, name collision APIs

dawdle-deer committed:
Merge branch 'master' of https://github.com/godotengine/godot into graphnode-revision

dawdle-deer committed:
fix various consistency issues

dawdle-deer committed:
Make new icons! and fix some issues

For now, GraphNode and GraphNodeIndexed have identical icons. In the future, changing this would be good! I just couldn't figure out a suitable icon (that didn't just look like a washing machine :P)

dawdle-deer committed:
fix documentation

dawdle-deer committed:
make a null check not error

dawdle-deer committed:
emit port disconnect signal properly

dawdle-deer committed:
fix crash on drag to disconnect, change default behaviour

dawdle-deer committed:
add some GraphEdit compatibility bindings

dawdle-deer committed:
revise documentation

dawdle-deer committed:
start assigning `_edit_lock_` to port_container automatically

dawdle-deer committed:
fix crashes on get_connections

dawdle-deer committed:
Add option to split snapping distances by axis

dawdle-deer committed:
add option to hide GraphNode title

dawdle-deer committed:
document previous commits

dawdle-deer committed:
make GraphNodeIndexed port layout account for hide_title

dawdle-deer committed:
fix title bar draw order, add a whole lot of theme item null checks

trying to find a crash on selection bug....

dawdle-deer committed:
add checks to fix a crash, small improvements

dawdle-deer committed:
add helpers for managing custom GraphNode titlebar nodes

dawdle-deer committed:
document GraphNode titlebar helpers

dawdle-deer committed:
fix GraphConnection queue_free outside of scene tree

dawdle-deer committed:
support undo/redo for GraphNodeIndexed slot removal, add helper

dawdle-deer committed:
fix script execution order, improve documentation

dawdle-deer committed:
fix execution order properly, remove debug comments

dawdle-deer committed:
improve port keyboard navigation and focus

nearing the finish line!!!!!!!!! :D

dawdle-deer committed:
finish keyboard usage

dawdle-deer committed:
further improve keyboard navigation

dawdle-deer committed:
update the docs

dawdle-deer commented:
The bugs are drying up and my todo list is complete, so I'm marking this as ready for review! I expect this will likely require more work to finish, and it still fails certain auto-run checks due to compatibility breakages (notably with GDExtension), but I think it's stable enough to stop calling it a draft ^^

dawdle-deer committed:
properly center ports, move icon offset to layout instead of draw

dawdle-deer committed:
move minimum graphport size out of custom_minimum_size

dawdle-deer committed:
Merge branch 'master' of https://github.com/dawdle-deer/godot into graphnode-revision

Koyper commented:
This is fantastic work! Thank you so much for this effort. If it's not too late to consider, I'd like to offer a few suggestions that would greatly expand the kinds of connections `GraphEdit` could create, making it easier to generate orthogonally connected flow chart/schematic diagrams without becoming too specialized. I have experience implementing a complete diagramming application like Visio.

**For the new `GraphPort`:**
 I think an additional enum specifying the vector direction of the connection attached to the port should be added, which is a different concept than the input/output flow of information now in the `PortDirection`. Maybe rename `PortDirection` to `PortFlowDirection`. Also add to it a `BIDIRECTIONAL` member which will inform the potential use of an arrowhead at a port. There are cases where a `GraphNode` symbol could have a `GraphPort` with an ambiguous connection vector start/end direction, so custom connection line routing code should be provided that direction. I've attached a couple of examples of symbols with ambiguous port vector directions where relying on the port relationship to the `Rect2` extents of the `GraphNode` to compute the direction vector would not work. In general, any port not on the rectangular perimeter of the `GraphNode` or in a corner will have an ambiguous vector direction.

```
enum PortFlowDirection {
	INPUT,
	OUTPUT,
	BIDIRECTIONAL,
	UNDIRECTED
};
enum PortConnectionDirection {
	NORTH,
	SOUTH,
	EAST,
	WEST,
	OMNIDIRECTIONAL
};
```
Now modify `GraphEdit::get_connection_line()` to pass the ports so a virtual method can create a line using the connection directions.
`	virtual PackedVector2Array get_connection_line(const Vector2 &p_from, const Vector2 &p_to, GraphPort &p_from_port, GraphPort &p_to_port) const;`

It would be nice to add a `text` property to `GraphPort` so ports can be named. This would allow automatic connection from/to labeling and `GraphNode` port naming.

**Connection Arrowheads**
Since you now have the ports provided with the `GraphEdit::connection_request` signal, we can obtain both port's `PortFlowDirection` and `PortConnectionDirection`, which is enough to determine if an arrowhead should be added and in what orientation.

It would be fantastic if you could actually implement arrowheads with this PR, but optional and could be done in a future PR. It's slightly non-intuitive to add the arrowheads to `GraphPort`, but I think that's the place to do it.

First we add a theme icon for the arrowhead to `GraphPort`. I've attached an example arrowhead `SVGTexture` resource.

Then in `GraphEdit` expose a `connection_arrowheads` property:
```
enum ConnectionArrowheads {
	ARROWHEAD_INPUT_OR_BIDIRECTIONAL,
	ARROWHEAD_BOTH_ENDS,
	ARROWHEAD_NONE
};

ConnectionArrowheads connection_arrowheads = ARROWHEAD_NONE;
```
Add the property to `GraphPort`, and this property determines if the arrowhead gets added in `NOTIFICATION_DRAW`, and it's oriented in accordance with `PortConnectionDirection`: 
```
GraphEdit::ConnectionArrowheads connection_arrowheads = graph_edit->connection_arrowheads;
```
**Misc**
When resizing a `GraphNode`, holding down the ALT key should lock the `GraphNode` aspect ratio.

I think connection colors should also be able to be set independently of their port color.

`SVGTexture` arrowhead icon:
[graph_port_arrowhead.tres.zip](https://github.com/user-attachments/files/21720663/graph_port_arrowhead.tres.zip)

<img width="1152" height="651" alt="Screenshot 2025-08-11 at 1 18 16 PM" src="https://github.com/user-attachments/assets/cea3fcbf-c2a1-4413-ae14-6c5054bb6367" />

Koyper commented:
> I agree that adding control over the direction connections should enter/exit the port would be a good idea. For flexibility, I think this might be best encoded as a float angle in degrees, rather than an enum, and it would be nice to make it a theme item.
> 
Seems to me it should be a property rather than a theme setting, because each port will have it calculated or determined individually, and there is no case where one theme setting would work for all ports.

One advantage of having the compass directions as enum choices, is you avoid having to compare floats to check if the angle is orthoganal, which would be error-prone. The compass enums could be set to ints 0, 90, 180 and 270 and cast to a float for the angle var, and the angle property could be exposed and used only when the direction is OMNIDIRECTIONAL.

Custom routing logic would use something like a lookup table given the start and end directions for all possible routings, so a `Dictionary` key would be something like `[NORTH, EAST]` so you'd want the directions as enums as a matter of convenience.
> Additionally, splitting it into multiple values would allow independent control over the direction of input, output, and undirected connections on a single port, which i can see being handy.
>
Interesting idea - might be too specialized since that could be done by the routing logic? Which brings up a thought about `GraphConnection` (for future PR with arrowheads  per your comment below) - I wondered if connections should store the direction obtained from the port, and then draw the arrowhead there? Then if you had different input/output directions the code might be simpler. Also if you delete a connection it automatically takes the arrowhead with it. So is there any other reason why we might want to store the start/end directions in `GraphConnection`?
> > Also add to it a `BIDIRECTIONAL` member
> 
> How does this differ from Undirected? For reference, an Undirected port allows connections from other ports regardless of their direction. Do you mean that this port should, when connected, assign itself an implicit direction based on that connection?
> 
Yes, I was uncertain what Undirected implied - does this mean it becomes the polar opposite of what you connect it to? If so, that's different from Bidirectional in terms of being what determines whether or not to use an arrowhead. If the intent is information flows in both directions, then maybe Bidirectional is a bit more precise?
> > It would be nice to add a `text` property to `GraphPort` so ports can be named.
> 
> Agreed! Though, figuring out the right layout behaviours for that text sounds like quite a task. I think this would warrant a whole followup PR. Plus, `GraphPort` being a node means it's extensible, so adding features like that can even be handled in script ^^
>
Here I was meaning the `text` property is just for the name of the port, and where to show that with a `Label` or something would be handled elsewhere.
> > The port type naming scheme via a `Dictionary` in `GraphEdit` seems like an unusual pattern in Godot that must have been in response to the old automatic port indexing architecture.
> 
> Quite unusual yeah. I believe it's being used somewhere in the visual shader editor (possibly for screen reader things?), though to be honest I don't know what it's even used for. Regardless, it's separate from naming the port, since it's naming the _type_ of a port.
>
Isn't the port name a string for whatever the type enum int is? I'll try to hunt down how that's used in visual shader editor, but it seems like the naming assignments would always be handled by the class implementing `GraphEdit` and would then do it by port if it had a name/text property. Maybe this PR shouldn't try to move that code up to `VisualShader`, but it would decouple things better if it were not in `GraphEdit`.

I also had a question about this PR's `type_colors` `Array` - should this be in `GraphEdit` at all per above? A parent class could manage the port type->color info. If it stays, it seems like it needs to be a `Dictionary` like `type_names` so the port type ints can be any set of values instead of only sequential.
> > It would be fantastic if you could actually implement arrowheads with this PR, but optional and could be done in a future PR.
> 
> I'm veto-ing this for this PR, but it sounds like a good feature for a followup PR.
> 
This is a good choice and can be done later, and easily if we lay the groundwork.
> This also makes me think it would be a good idea to split a port's theme icon to allow icons to be assigned to input, output, and undirected ports at the theme level.
>
Yes, and then the icon draws per the `PortFlowDirection` - although this could be added in a future PR too.

Thanks again for your work on this!

dawdle-deer commented:
> Seems to me it should be a property rather than a theme setting, because each port will have it calculated or determined individually.

Typically from what I've seen, graphs have consistent directions of travel for better readability - inputs and outputs flow from left to right, top to bottom, etc. Obviously, I want to make this system flexible enough to support graphs that break this, but I'd prefer to make the most commonly-traveled path the one with least resistance. Hence, encode direction as a theme item to let users with consistently-directed ports "fire and forget" their port angles, but still allow users with more complex graphs override these values on a per-port basis just fine ^^

I agree that it makes orthogonality checks slightly more difficult, but most projects don't need to compare the precise angle their ports provide to connections in script. For the ones that do, something simple like `floor(port1.conn_direction) == floor(port2.conn_direction)` should do the trick.

> I wondered if connections should store the direction obtained from the port, and then draw the arrowhead there?

Yep, drawing the arrowhead using the connection makes complete sense to me! There's already logic in the drawing of the connection that accesses a port's direction for this purpose.

> Yes, I was uncertain what Undirected implied - does this mean it becomes the polar opposite of what you connect it to?

Alright, so here's how the directions work:
- A port marked as `INPUT` may make connections to an `OUTPUT` or `UNDIRECTED` port. It may not make connections to another `INPUT` port.
- A port marked as `OUTPUT` may make connections to an `INPUT` or `UNDIRECTED` port. It may not make connections to another `OUTPUT` port.
- A port marked as `UNDIRECTED` may make connections to any port, regardless of direction. An `UNDIRECTED` port may be connected to both `INPUT` and `OUTPUT` ports (even at the same time!). Think of it as a "this port ignores directions" option - it is non-directed.

There's room here to add some behaviour that, like you assumed with `BIDIRECTIONAL`, makes a port undirected by default, but imply its direction from its first connection to another directed port (if it has a connection). I haven't run into this yet, as an `UNDIRECTED` port that allows more than one connection (IE is not `exclusive`) is something of a niche use case, but it'd certainly be useful in that case.

> Here I was meaning the text property is just for the name of the port, and where to show that with a Label or something would be handled elsewhere.

True, but I think both the property and the behaviour that displays it with a `Label` should be added at the same time, as adding an unused property to the engine is kinda bad practice.

The hard part of that whole task would be designing the layout behaviour for that label - lots of tricky design decisions in making it flexible for custom text layouts but still friendly to use with little-to-no input for the majority of projects. If someone wants to test that out, an intermediate step some folks take is writing a GDExtension to experiment with things - I think it might be a good approach.

> I also had a question about this PR's type_colors Array - should this be in GraphEdit at all per above?

If it's stored in either the `GraphNode` or `GraphPort`, it has to be serialized in _every_ `GraphNode`/`GraphPort`, so this saves a great deal on storage. But, more importantly, I found it was pretty terrible practice for port types and type colors to be determined locally within the same graph. If two ports in the same graph have the same port type number, they should _be_ the same type semantically as well - type `1` shouldn't be, say, `int` in one place and `Oregano` in another. Similarly, if two ports have the same icon and color, it indicates they're the same type. There's exceptions, like if `1` is `int` and `12` is `enum`, it's fine to make them both the same shade of cyan (so long as you call `GraphEdit.add_valid_connection_type` to make them interchangeable to the user). But from a UX point of view, the user should be presented with a consistent, uniform color palette that signals "light blue can be connected to light blue", and making the port type color palette local to either `GraphNode` or `GraphPort` is just asking for projects to break this pattern for very little gain. I find making this a graph-global property discourages bad port-typing practice across the board.

This logic probably also applies to type names, but for this PR, I'm likely not going to touch that property lol.

Anyway, you're definitely right, it should be a typed dictionary instead of an array :P

Ty for the feedback!

dawdle-deer committed:
Add imply_direction to GraphPort

dawdle-deer committed:
convert port type colors from array to dictionary

dawdle-deer committed:
catch errors slightly

dawdle-deer committed:
add connection_angle theme items to GraphPort

dawdle-deer committed:
hide GraphPort direction if it's implied

dawdle-deer committed:
update docs, fix small issues

dawdle-deer commented:
@Koyper based on your suggestions, I've done the following:
- Added `GraphPort.imply_direction`, which will make a port decide its direction based on the first connection to an input or output port
- Changed `GraphEdit.type_colors` to a `Dictionary[int, Color]`
- Added `GraphPort.get_connection_angle`, which is driven by some theme items and used to control connection line angles.

AThousandShips commented:
> You will also need to squash your commits at some point.

This is not an issue until it's fully reviewed and ready to merge, until then it's up to you what works best for your workflow

dawdle-deer committed:
add methods to hide GraphEdit scrollbars

dawdle-deer committed:
fix connection direction slightly

dawdle-deer committed:
clean up + fix port interaction

this should also help properly support standalone ports

dawdle-deer committed:
document scrollbar_hidden

dawdle-deer commented:
> have you thought about the idea of having an `HGraphPort` and `VGraphPort`?

The H and V hotzone constants refer to width and height, not horizontal and vertical port orientation. Ports will still need width and height settings in either orientation, so that change would not simplify those properties.

In any case, I think it would probably be best to stick to the base `GraphPort` class in-engine. Ports are usually more-or-less square, and shouldn't be used as layout containers like `HBoxContainer`.

Separately, I _do_ see a use case for a `VGraphNodeIndexed`/`HGraphNodeIndexed` split. That's worth considering for another PR ^^

dawdle-deer commented:
> Why doesn't `GraphEdit` use a `Viewport` with a `Camera2D`? 

No idea tbh! It seems like a redundant solution to a solved problem, but there could be reasons.

> Is it too early to report issues with `VisualShader`?

Nope, please fire away, I want to squash regressions there asap.

> I think GraphElements without GraphPorts should still draw in the minimap. 

Agreed! Plus, I was thinking about making `GraphPort`s with no parent be drawn in the minimap, so drawing all `GraphElement`s would also accomplish that. 

> It would be most useful if an imply direction port were also automatically set to `Output` if it's the first port selected when dragging in a connector. 

That's an excellent idea! Yeah, If both ports in a connection are implied, I think it makes sense to infer that from user input.

> I'd like to propose a new `GraphEdit` method `get_tool_menu_control()` that returns the tool internal `Control` node, so you can move or reparent it or add additional items to it.

Also a good idea! Consider it done.

Koyper commented:
> Hey wait, isn't this just `GraphEdit.get_menu_hbox`?

Doy! I was looking for a `Control` return method. Proposal resolved!

dawdle-deer committed:
make ports imply and enforce direction better

dawdle-deer committed:
make minimap draw GraphElement universally

dawdle-deer committed:
fix free-floating GraphPort draw and connections

dawdle-deer committed:
Merge branch 'master' of https://github.com/dawdle-deer/godot into graphnode-revision

# Conflicts:
#	editor/themes/editor_theme_manager.cpp

Naros commented:
Hi @Koyper, great feedback, but I don't concur with the rename of `GraphNode`.

With respect, I understand your intent here with aligning with other GUI elements, but at the same time, I do wonder if eliminating the idea of "node" in the class name could be more confusing for users. In visual graphs, it's commonly accepted that you have nodes, connections, and connections are between pins/ports on nodes. 

So if anything, I think using `GraphNodeContainer` as the stem makes more sense from a layman's understanding of the class structure rather than `GraphPortContainer`, but that's just my $0.02.

dawdle-deer commented:
> The editor crashes when changing Vector Compose `Vector3` to `Vector4` or `Vector2`.

Hmm, this works fine on my end. Could I get some more info about how you're producing this, or the call stack on crash?

> Shouldn't [`GraphFrame`] act like a `Control` when disabled?

I'll look into this, but I believe this down to some hard-coded `Control` behaviour that checks to see if a node's parent is a `Container`, which would be a separate PR to address.

> On drag, snap `position + pivot_offset` to the grid, not just `position`.
> Also, resizing a graph node should resize around the `pivot_offset`, not `position`.

Will do!

> It would be super useful to be able to disable GraphEdit so that nothing is editable and zoom/pan are locked.

Sure! I can add `view_locked` for this purpose. 

> Is there any reason why all other `Control` children of `GraphEdit` shouldn't act like they are inside a `GraphElement`?

`GraphEdit`'s current behaviour is fairly dependent on properties and methods that `GraphElement` exposes. This opens up a question I think is very useful: what exactly _is_ a `GraphElement`? I believe this UI element is more accurately described as a `ResizeableContainer`, with extra behaviour thrown in to let it be selected, dragged, and themed with a custom minimap panel. This leads to... well, a much broader point.

To be honest, having rewritten this system to this point, I believe there's still one major step towards integrating this system with the engine as a whole and removing the experimental label. This would require yet another major reimplementation, building on top of this PR. Specifically, I believe the following would bring this system much more in line with Godot UI's design up to this point:
- Make `GraphConnection` a node inheriting `Line2D`, and rename it to something more general like `ConnectionLine2D`.
- Rather than inheriting from `GraphElement`, as it is not a `Container`, make `GraphPort` inherit from a new node called `DragHandle`. Additionally, there's an argument to be made here that `GraphPort` should work completely in isolation, no `GraphEdit` required.
- As stated above, `GraphElement` should be refactored into `ResizeableContainer`. This would require selection logic to be extended to work with any `Control`, likely handled by `GraphEdit`, and while the details of this in particular sound _very_ sticky, it would make this system much more powerful.
- Look into making `GraphEdit` inherit from `ViewportContainer`, and perhaps renaming it to something like `GraphView`.

That said, I believe that's not only completely out of scope for this PR, it would be best implemented as _several_ PRs. I think `DragHandle` is something that should be introduced in a dedicated PR, if its existence is warranted (I've implemented this component myself in GDScript a couple times for VNML, so I can vouch for its usefulness). These are potential improvements, best explored in future work.
